src/elf_parser/mod.rs

```diff
@@ -0,0 +1,616 @@
+//! Dependency-less 64 bit ELF parser
+
+pub mod consts;
+pub mod types;
+
+use std::ops::Range;
+use {crate::ebpf, consts::*, types::*};
+
+const EXPECTED_PROGRAM_HEADERS: [(u32, u32, u64); 3] = [
+    (PT_LOAD, PF_R | PF_X, ebpf::MM_PROGRAM_START),
+    (PT_GNU_STACK, PF_R | PF_W, ebpf::MM_STACK_START),
+    (PT_NULL, PF_R | PF_W, ebpf::MM_HEAP_START),
+];
+const SECTION_COUNT_MAXIMUM: usize = 16;
+const SECTION_NAME_LENGTH_MAXIMUM: usize = 16;
+const SYMBOL_NAME_LENGTH_MAXIMUM: usize = 64;
+
+/// Error definitions
+#[derive(Debug, PartialEq, Eq, thiserror::Error)]
+pub enum ElfParserError {
+    /// ELF file header is inconsistent or unsupported
+    #[error("invalid file header")]
+    InvalidFileHeader,
+    /// Program header is inconsistent or unsupported
+    #[error("invalid program header")]
+    InvalidProgramHeader,
+    /// Section header is inconsistent or unsupported
+    #[error("invalid section header")]
+    InvalidSectionHeader,
+    /// Section or symbol name is not UTF8 or too long
+    #[error("invalid string")]
+    InvalidString,
+    /// An index or memory range does exeed its boundaries
+    #[error("value out of bounds")]
+    OutOfBounds,
+    /// The size isn't valid
+    #[error("invalid size")]
+    InvalidSize,
+    /// Headers, tables or sections do overlap in the file
+    #[error("values overlap")]
+    Overlap,
+    /// Sections are not sorted in ascending order
+    #[error("sections not in ascending order")]
+    SectionNotInOrder,
+    /// No section name string table present in the file
+    #[error("no section name string table found")]
+    NoSectionNameStringTable,
+    /// Invalid .dynamic section table
+    #[error("invalid dynamic section table")]
+    InvalidDynamicSectionTable,
+    /// Invalid relocation table
+    #[error("invalid relocation table")]
+    InvalidRelocationTable,
+    /// Invalid alignment
+    #[error("invalid alignment")]
+    InvalidAlignment,
+    /// No string table
+    #[error("no string table")]
+    NoStringTable,
+    /// No dynamic string table
+    #[error("no dynamic string table")]
+    NoDynamicStringTable,
+}
+
+fn check_that_there_is_no_overlap(
+    range_a: &std::ops::Range<usize>,
+    range_b: &std::ops::Range<usize>,
+) -> Result<(), ElfParserError> {
+    if range_a.end <= range_b.start || range_b.end <= range_a.start {
+        Ok(())
+    } else {
+        Err(ElfParserError::Overlap)
+    }
+}
+
+/// The parsed structure of an ELF file
+pub struct Elf64<'a> {
+    elf_bytes: &'a [u8],
+    file_header: &'a Elf64Ehdr,
+    program_header_table: &'a [Elf64Phdr],
+    section_header_table: &'a [Elf64Shdr],
+    section_names_section_header: Option<&'a Elf64Shdr>,
+    text_section_header: Option<&'a Elf64Shdr>,
```

Comment:
Why does this struct contain `text_section_header` and `readonly_data_section_header`?  Seems these members are never used?

Reply:
Good catch, removed!